### PR TITLE
Parallel transport for invariant metric

### PR DESCRIPTION
**Pull request by nguigs**

### CODE REVIEW

**Comment by nguigs** [comment]
No idea what happened with the notebook that failed a test

**Comment by nguigs** [comment]
No idea what happened with the notebook that failed a test

File: geomstats/geometry/invariant_metric.py [hunk 1]
```diff
@@ -652,6 +651,117 @@ def log(self, point, base_point, n_steps=15, step='rk4',
...
 
+    def parallel_transport(
+            self, tangent_vec_a, tangent_vec_b, base_point, n_steps=10,
+            step='rk4', return_endpoint=False):
+        r"""Compute the parallel transport of a tangent vec along a geodesic.
+
+        Approximate solution for the parallel transport of a tangent vector a
+        along the geodesic defined by :math: `t \mapsto exp_(base_point)(t*
+        tangent_vec_b)`. The parallel transport equation is written entirely
+        in the Lie algbera and solved with an integration scheme.
```
Comment (ninamiolane) [inline geomstats/geometry/invariant_metric.py:662]: NIT: algebra (not algbera)

File: geomstats/geometry/invariant_metric.py [hunk 2]
```diff
@@ -652,6 +651,117 @@ def log(self, point, base_point, n_steps=15, step='rk4',
                 point, base_point, n_steps=n_steps, step=step,
                 verbose=verbose, max_iter=max_iter, tol=tol), base_point)
 
+    def parallel_transport(
```
Comment (ninamiolane) [inline geomstats/geometry/invariant_metric.py:654]: Nice!! 🎉

File: geomstats/geometry/invariant_metric.py [hunk 3]
```diff
@@ -652,6 +651,117 @@ def log(self, point, base_point, n_steps=15, step='rk4',
...
+        initial_state = gs.stack([
+            base_point, left_angular_vel_b, left_angular_vel_a])
+        flow = integrate(
+            acceleration, initial_state, n_steps=n_steps, step=step)
+        gamma, _, zeta_t = flow[-1]
+        transported = group.tangent_translation_map(
+            gamma, left_or_right=self.left_or_right, inverse=False)(zeta_t)
+        return (transported, gamma) if return_endpoint else transported
+
+    def geodesic_equation(self, state, _time):
```
Comment (ninamiolane) [inline geomstats/geometry/invariant_metric.py:732]: Do you have a reference for this (in addition to the formula?)?

File: geomstats/geometry/invariant_metric.py [hunk 4]
```diff
@@ -1086,12 +1196,12 @@ def inner_product(self, tangent_vec_a, tangent_vec_b, base_point=None):
...
     def parallel_transport(self, tangent_vec_a, tangent_vec_b, base_point):
-        r"""Compute the parallel transport of a tangent vector.
+        r"""Compute the parallel transport of a tangent vec along a geodesic.
 
         Closed-form solution for the parallel transport of a tangent vector a
         along the geodesic defined by :math: `t \mapsto exp_(base_point)(t*
-        tangent_vec_b)`. As the special Euclidean group endowed with its
-        canonical left-invariant metric is a symmetric space, parallel
+        tangent_vec_b)`. As a Lie group endowed with its
+        canonical bi-invariant metric is a symmetric space, parallel
```
Comment (ninamiolane) [inline geomstats/geometry/invariant_metric.py:1204]: Not all Lie groups have a bi-invariant metric: what do you mean here?

File: geomstats/geometry/special_euclidean.py [hunk 5]
```diff
@@ -185,8 +185,13 @@ def inverse(cls, point):
 
         Parameters
         ----------
-        point : array-like, shape=[..., n, n]
+        point : array-like, shape=[..., n + 1, n + 1]
```
Comment (ninamiolane) [inline geomstats/geometry/special_euclidean.py:188]: well spotted! 🙏

File: geomstats/integrator.py [hunk 6]
```diff
@@ -88,7 +88,8 @@ def rk4_step(force, state, time, dt):
     state : array-like, shape=[2, dim]
         State at time t, corresponds to position and velocity variables at
         time t.
-    time :
+    time ; float
```
Comment (ninamiolane) [inline geomstats/integrator.py:91]: NIT: ; -> :

File: tests/tests_geomstats/test_invariant_metric.py [hunk 7]
```diff
@@ -776,3 +776,34 @@ def test_dist_pairwise_parallel(self):
         belongs = Matrices(n_samples, n_samples).belongs(result)
         self.assertTrue(is_sym)
         self.assertTrue(belongs)
+
+    def test_integrated_parallel_transport(self):
```
Comment (ninamiolane) [inline tests/tests_geomstats/test_invariant_metric.py:780]: Maybe add a test for the code on the geodesic equation too?

**Review by ninamiolane — APPROVED** [review]
Awesome, thank you very much for adding these! I've put a few comments + unit tests need to pass.

File: geomstats/geometry/invariant_metric.py (hunk 4 above)
Comment (nguigs) [reply geomstats/geometry/invariant_metric.py:1204]: As we are in the bi-invariant class, I was assuming the group considered has a bi-invariant metric! But I can add *compact* Lie group

File: tests/tests_geomstats/test_invariant_metric.py (hunk 7 above)
Comment (nguigs) [reply tests/tests_geomstats/test_invariant_metric.py:780]: This is tested by verifying that the end_point of PT coincides with exp

Comment (nguigs) [reply tests/tests_geomstats/test_invariant_metric.py:780]: I don't think that testing only the shape of the output is relevant, and can't think of an expected value for `geodesic_equation` alone

**Pull request merged**